libraries/maplibre-compose/src/main/kotlin/io/element/android/libraries/maplibre/compose/CameraPositionState.kt

```diff
@@ -0,0 +1,189 @@
+     * panning, zooming, or rotation.
+     */
+    public var isMoving: Boolean by mutableStateOf(false)
+        internal set
+
+    /**
+     * The reason for the start of the most recent camera moment, or
+     * [CameraMoveStartedReason.NO_MOVEMENT_YET] if the camera hasn't moved yet or
+     * [CameraMoveStartedReason.UNKNOWN] if an unknown constant is received from the Maps SDK.
+     */
+    public var cameraMoveStartedReason: CameraMoveStartedReason by mutableStateOf(
+        CameraMoveStartedReason.NO_MOVEMENT_YET
+    )
+        internal set
+
+    /**
+     * Returns the current [Projection] to be used for converting between screen
+     * coordinates and lat/lng.
+     */
+    public val projection: Projection?
+        get() = map?.projection
+
+    /**
+     * Local source of truth for the current camera position.
+     * While [map] is non-null this reflects the current position of [map] as it changes.
+     * While [map] is null it reflects the last known map position, or the last value set by
+     * explicitly setting [position].
+     */
+    internal var rawPosition by mutableStateOf(position)
+
+    /**
+     * Current position of the camera on the map.
+     */
+    public var position: CameraPosition
+        get() = rawPosition
+        set(value) {
+            synchronized(lock) {
+                val map = map
+                if (map == null) {
+                    rawPosition = value
+                } else {
+                    map.moveCamera(CameraUpdateFactory.newCameraPosition(value))
+                }
+            }
+        }
+
+    /**
+     * Local source of truth for the current camera mode.
+     * While [map] is non-null this reflects the current camera mode as it changes.
+     * While [map] is null it reflects the last known camera mode, or the last value set by
+     * explicitly setting [cameraMode].
+     */
+    internal var rawCameraMode by mutableStateOf(cameraMode)
+
+    /**
+     * Current tracking mode of the camera.
+     */
+    public var cameraMode: CameraMode
+        get() = rawCameraMode
+        set(value) {
+            synchronized(lock) {
+                val map = map
+                if (map == null) {
+                    rawCameraMode = value
+                } else {
+                    map.locationComponent.cameraMode = value.toInternal()
+                }
+            }
+        }
+
+    /**
+     * The user's last available location.
+     */
+    public var location: Location? by mutableStateOf(null)
+        internal set
+
+    // Used to perform side effects thread-safely.
+    // Guards all mutable properties that are not `by mutableStateOf`.
+    private val lock = Unit
+
+    // The map currently associated with this CameraPositionState.
+    // Guarded by `lock`.
+    private var map: MapboxMap? by mutableStateOf(null)
+
+    // The current map is set and cleared by side effect.
```

Comment:
It means that we can't just pass the map object in the state class when we build it and keep it as a val, because the `CameraPositionState`'s lifetime can be greater than that of the map.
So whenever the map is destroyed and a new instance built, we will update the state object with a reference to the new instance of the map. The code that does this "updating" side effect is in `MapUpdater.kt`